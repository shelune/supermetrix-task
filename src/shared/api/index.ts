type ApiPostType = "status";

export type ApiPost = {
  id: string;
  from_name: string;
  from_id: string;
  message: string;
  type: ApiPostType;
  created_time: string;
};

export type ApiPostResponse = {
  meta: {
    request_id: string;
  };
  data: {
    page: number;
    posts: ApiPost[];
  };
};

export type Post = {
  id: string;
  authorName: string;
  authorId: string;
  message: string;
  type: ApiPostType;
  createdTime: Date;
};

export type ApiRegisterResponse = {
  meta: {
    request_id: string;
  };
  data: {
    client_id: string;
    email: string;
    sl_token: string;
  };
};
