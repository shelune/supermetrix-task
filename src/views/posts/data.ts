import { useParams } from "react-router-dom";
import axios from "axios";
import { ComponentProps, useCallback, useEffect, useState } from "react";
import type { ApiErrorResponse, ApiPostResponse, Post } from "../../shared/api";
import type { PostsView } from "./posts";

import { postMapper } from "../../shared/utils/mapper";
import { getCookie } from "../../shared/utils/cookie";

export const MAX_PAGE = 10;

export const useData = (): ComponentProps<typeof PostsView> => {
  const [posts, setPosts] = useState<Post[]>([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState("");
  const [page, setPage] = useState(1);
  const [activeUser, setActiveUser] = useState("");
  const params = useParams();

  const getData = useCallback(async () => {
    setLoading(true);
    try {
      const response = await axios.get<ApiPostResponse>(
        "https://api.supermetrics.com/assignment/posts",
        {
          params: {
            sl_token: getCookie("sl_token"),
            page,
          },
        },
      );
      const mapped = response.data.data.posts.map(postMapper);
      setPosts(posts.concat(mapped));
      setError("");
    } catch (err) {
      if (axios.isAxiosError(err) && err.response?.data) {
        const errorResponse = err.response.data as ApiErrorResponse;
        setError(errorResponse.error.message);
      } else {
        // eslint-disable-next-line no-console
        console.log("unknown error when getting posts", err);
      }
    }
    setLoading(false);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [page]);

  useEffect(() => {
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    getData();
  }, [getData, page]);

  useEffect(() => {
    if (params.userId) {
      setActiveUser(params.userId);
    } else {
      setActiveUser("");
    }
  }, [params.userId]);

  const loadMore = useCallback(() => {
    if (page + 1 > MAX_PAGE) {
      return;
    }
    setPage(page + 1);
  }, [page]);

  return {
    posts,
    currentPage: page,
    activeUser,
    error,
    loading,
    loadMore,
  };
};
