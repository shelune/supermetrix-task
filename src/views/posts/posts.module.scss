@import "../../styles/utils/_index.scss";

.container {
  padding: 24px;
  display: grid;
  grid-template-columns: 1fr 2fr;
  gap: 16px;
  align-items: start;
}

.error {
  color: $red-error;
  text-align: center;
}
