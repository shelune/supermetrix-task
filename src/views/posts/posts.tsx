import React, { FC } from "react";
import { Post } from "../../shared/api";
import { PostColumn } from "./post-column/post-column";

import css from "./posts.module.scss";
import { UserColumn } from "./user-column/user-column";

export type UserData = {
  name: string;
  userId: string;
  postCount: number;
};

type Props = {
  posts: Post[];
  currentPage: number;
  activeUser: string;
  error: string;
  loading: boolean;
  loadMore: () => void;
};

export const PostsView: FC<Props> = ({
  posts,
  currentPage,
  activeUser,
  error,
  loading,
  loadMore,
}) => (
  <>
    {error && <div className={css.error}>{error}</div>}
    <div className={css.container}>
      <UserColumn posts={posts} activeUser={activeUser} loading={loading} />
      <PostColumn
        posts={posts}
        currentPage={currentPage}
        loading={loading}
        loadMore={loadMore}
      />
    </div>
  </>
);
