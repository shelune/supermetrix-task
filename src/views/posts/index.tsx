import React, { FC, memo } from "react";
import { useData } from "./data";
import { PostsView as MainComponent } from "./posts";

export const PostsView: FC = memo(
  () => {
    const props = useData();
    // eslint-disable-next-line react/jsx-props-no-spreading
    return <MainComponent {...props} />;
  },
  () => true,
);
