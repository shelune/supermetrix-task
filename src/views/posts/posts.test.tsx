import axios from "axios";
import React, { ReactNode } from "react";
import { act, fireEvent, render } from "@testing-library/react";
import { PostsView } from "./posts";
import { mockPosts } from "../../shared/__mock__/posts";
import { Post } from "../../shared/api";

jest.mock("./post-column/post-column.tsx", () => ({
  PostColumn: ({ posts }: { posts: Post[] }) => (
    // eslint-disable-next-line react/destructuring-assignment
    <div>{`Post Column: ${posts.length} posts`}</div>
  ),
}));

jest.mock("./user-column/user-column.tsx", () => ({
  UserColumn: () => (
    // eslint-disable-next-line react/destructuring-assignment
    <div>User Column</div>
  ),
}));

describe("PostsView", () => {
  it("should render", () => {
    const { container } = render(
      <PostsView
        posts={mockPosts}
        currentPage={1}
        activeUser=""
        error=""
        loading={false}
        loadMore={jest.fn()}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
