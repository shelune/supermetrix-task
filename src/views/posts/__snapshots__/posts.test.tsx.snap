// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsView should render 1`] = `
<div>
  <div
    class="container"
  >
    <div>
      User Column
    </div>
    <div>
      Post Column: 100 posts
    </div>
  </div>
</div>
`;
